In [1]:
import pandas as pd

df_Standardscalar = pd.DataFrame({
    'Normalization Technique': ['Standardscalar', 'Standardscalar', 'Standardscalar','Standardscalar'],
    'Algorithm': ['K-Means', 'Agglomerative', 'DBSCAN','Guassian_Mixture_Model'],
    'No.Of Clusters': ['3','3','0.3','4'],
    'Silhouette Score': [0.28605038759583024, 0.26797203261079094, -0.3636105417264654, 0.0895031941931934],
    'Calinski-Harabasz Index': [1391.8542662333207, 1258.1943544223623, 4.075530344037714, 2956.121046372958],
    'Davies-Bouldin Index': [1.2877580801088788, 1.2993235430484735, 1.7659704926628443, 2.132704317187362]
})

df_minmax = pd.DataFrame({
    'Normalization Technique': ['Minmax', 'Minmax', 'Minmax','Minmax'],
    'Algorithm': ['K-Means', 'Agglomerative', 'DBSCAN','Guassian_Mixture_Model'],
    'No.Of Clusters': ['3','3','0.3','8'],
    'Silhouette Score': [0.2932325850872667, 0.2714059233809167, 0.003926820171041113,-0.15861655325661503],
    'Calinski-Harabasz Index': [1586.9282389231478, 1349.7520769147925, 24.538256499755537,2193.199284048159],
    'Davies-Bouldin Index': [1.2293117099049238, 1.2364289451780115, 1.233450013513498,7.4583285936553425]
})

df_log = pd.DataFrame({
    'Normalization Technique': ['Log', 'Log', 'Log','Log'],
    'Algorithm': ['K-Means', 'Agglomerative', 'DBSCAN','Guassian_Mixture_Model'],
    'No.Of Clusters': ['3','3','0.3','9'],
    'Silhouette Score': [0.30471357555611095, 0.24772039344825694, -0.3755188461522972,-0.24244006853848216],
    'Calinski-Harabasz Index': [1655.2673373853254, 1420.65529966141, 12.293935073395863,1127.2123383495598],
    'Davies-Bouldin Index': [1.3240327126662048, 1.3867421601443108, 1.049946366346552,4.580933212530849]
})

df_sqrt = pd.DataFrame({
    'Normalization Technique': ['Sqrt', 'Sqrt', 'Sqrt','Sqrt'],
    'Algorithm': ['K-Means', 'Agglomerative', 'DBSCAN','Guassian_Mixture_Model'],
    'No.Of Clusters': ['3','3','0.3','9'],
    'Silhouette Score': [0.6376133571726594, 0.657742204384634, -0.6152871547593055,0.18153634226695156],
    'Calinski-Harabasz Index': [15160.541299442248, 14334.219873148091, 29.40407865175737,5509.430499485822],
    'Davies-Bouldin Index': [0.5304374515938449, 0.5142035923598317, 1.0123464306140635,1.6308770418457348]
})

df_Yeojohnson = pd.DataFrame({
    'Normalization Technique': ['Yeojohnson', 'Yeojohnson', 'Yeojohnson','Yeojohnson'],
    'Algorithm': ['K-Means', 'Agglomerative', 'DBSCAN','Guassian_Mixture_Model'],
    'No.Of Clusters': ['3','3','0.3','9'],
    'Silhouette Score': [0.6239342306071544, 0.6239342306071544, -0.3693726715419584,0.07723558526217594],
    'Calinski-Harabasz Index': [10809.816975077965, 10809.816975077965, 0.2668273485391091,5292.0543733317845],
    'Davies-Bouldin Index': [0.46058011126145315, 0.46058011126145315, 4.252341123785477,1.287347585372836]
})

# Concatenate the new DataFrame with the existing one
df_all = pd.concat([df_Standardscalar,df_minmax, df_log,df_sqrt,df_Yeojohnson], ignore_index=True)

# Display the updated DataFrame
df_all

Normalization Technique               Algorithm No.Of Clusters  \
0           Standardscalar                 K-Means              3   
1           Standardscalar           Agglomerative              3   
2           Standardscalar                  DBSCAN            0.3   
3           Standardscalar  Guassian_Mixture_Model              4   
4                   Minmax                 K-Means              3   
5                   Minmax           Agglomerative              3   
6                   Minmax                  DBSCAN            0.3   
7                   Minmax  Guassian_Mixture_Model              8   
8                      Log                 K-Means              3   
9                      Log           Agglomerative              3   
10                     Log                  DBSCAN            0.3   
11                     Log  Guassian_Mixture_Model              9   
12                    Sqrt                 K-Means              3   
13                    Sqrt           Agglomerative              3   
14                    Sqrt                  DBSCAN            0.3   
15                    Sqrt  Guassian_Mixture_Model              9   
16              Yeojohnson                 K-Means              3   
17              Yeojohnson           Agglomerative              3   
18              Yeojohnson                  DBSCAN            0.3   
19              Yeojohnson  Guassian_Mixture_Model              9   

    Silhouette Score  Calinski-Harabasz Index  Davies-Bouldin Index  
0           0.286050              1391.854266              1.287758  
1           0.267972              1258.194354              1.299324  
2          -0.363611                 4.075530              1.765970  
3           0.089503              2956.121046              2.132704  
4           0.293233              1586.928239              1.229312  
5           0.271406              1349.752077              1.236429  
6           0.003927                24.538256              1.233450  
7          -0.158617              2193.199284              7.458329  
8           0.304714              1655.267337              1.324033  
9           0.247720              1420.655300              1.386742  
10         -0.375519                12.293935              1.049946  
11         -0.242440              1127.212338              4.580933  
12          0.637613             15160.541299              0.530437  
13          0.657742             14334.219873              0.514204  
14         -0.615287                29.404079              1.012346  
15          0.181536              5509.430499              1.630877  
16          0.623934             10809.816975              0.460580  
17          0.623934             10809.816975              0.460580  
18         -0.369373                 0.266827              4.252341  
19          0.077236              5292.054373              1.287348

In [2]:
df = pd.DataFrame({
    'Normalization Technique': ['Standardscalar', 'Standardscalar', 'Standardscalar', 'Standardscalar', 'Minmax', 'Minmax', 'Minmax', 'Minmax', 'Log', 'Log', 'Log', 'Log', 'Sqrt', 'Sqrt', 'Sqrt', 'Sqrt', 'YeoJohnson', 'YeoJohnson', 'YeoJohnson', 'YeoJohnson'],
    'Algorithm': ['K-Means', 'Agglomerative', 'DBSCAN', 'Guassian_Mixture_Model', 'K-Means', 'Agglomerative', 'DBSCAN', 'Guassian_Mixture_Model', 'K-Means', 'Agglomerative', 'DBSCAN', 'Guassian_Mixture_Model','K-Means','Agglomerative','DBSCAN','Guassian_Mixture_Model','K-Means','Agglomerative','DBSCAN','Guassian_Mixture_Model'],
    # Add your actual scores here
    'Silhouette Score': [0.28605038759583024, 0.26797203261079094, -0.3636105417264654,0.0895031941931934, 0.2932331586923892, 0.2714059233809167, 0.003926820171041113,-0.15861655325661503, 0.30471357555611095, 0.24772039344825694, -0.3755188461522972,-0.24244006853848216, 0.6376133571726594, 0.657742204384634, -0.6152871547593055,0.18153634226695156, 0.6239342306071544, 0.6239342306071544, -0.3693726715419584,0.07723558526217594],
    # Add your actual scores here
    'Calinski-Harabasz Index': [1391.8542662333207, 1258.1943544223623, 4.075530344037714,2956.121046372958, 1586.9282389231478, 1349.7520769147925, 24.538256499755537,2193.199284048159, 1655.2673373853254, 1420.65529966141, 12.293935073395863,1127.2123383495598, 15160.541299442248, 14334.219873148091, 29.40407865175737,5509.430499485822, 10809.816975077965, 10809.816975077965, 0.2668273485391091,5292.0543733317845],
    # Add your actual scores here
    'Davies-Bouldin Index': [1.2877580801088788, 1.2993235430484735, 1.7659704926628443,2.132704317187362, 1.2293117099049238, 1.2364289451780115, 1.2334500135134982,7.4583285936553425, 1.3240327126662048, 1.3867421601443108, 1.0499463663465522,4.580933212530849, 0.5304374515938449, 0.5142035923598317, 1.0123464306140635,1.6308770418457348, 0.46058011126145315, 0.46058011126145315 ,4.252341123785477,1.287347585372836]
})

# Find the algorithm with the highest Silhouette Score and Calinski-Harabasz Index and the lowest Davies-Bouldin Index
best_silhouette_score = df[df['Silhouette Score'] == df['Silhouette Score'].max()]
best_calinski_harabasz_index = df[df['Calinski-Harabasz Index'] == df['Calinski-Harabasz Index'].max()]
best_davies_bouldin_index = df[df['Davies-Bouldin Index'] == df['Davies-Bouldin Index'].min()]

print("Best Silhouette Score:\n", best_silhouette_score)
print("\nBest Calinski-Harabasz Index:\n", best_calinski_harabasz_index)
print("\nBest Davies-Bouldin Index:\n", best_davies_bouldin_index)


Best Silhouette Score:
    Normalization Technique      Algorithm  Silhouette Score  \
13                    Sqrt  Agglomerative          0.657742   

    Calinski-Harabasz Index  Davies-Bouldin Index  
13             14334.219873              0.514204  

Best Calinski-Harabasz Index:
    Normalization Technique Algorithm  Silhouette Score  \
12                    Sqrt   K-Means          0.637613   

    Calinski-Harabasz Index  Davies-Bouldin Index  
12             15160.541299              0.530437  

Best Davies-Bouldin Index:
    Normalization Technique      Algorithm  Silhouette Score  \
16              YeoJohnson        K-Means          0.623934   
17              YeoJohnson  Agglomerative          0.623934   

    Calinski-Harabasz Index  Davies-Bouldin Index  
16             10809.816975               0.46058  
17             10809.816975               0.46058  


# The best Silhouette Score of 0.657742 is achieved by the "Agglomerative algorithm" with the "Sqrt" normalization technique.
# The best Calinski-Harabasz Index of 15160.541299 is achieved by the "K-Means algorithm" with the "Sqrt" normalization technique.
# The best Davies-Bouldin Index of "0.46058" is achieved by both the "K-Means" and "Agglomerative algorithms" with the YeoJohnson normalization technique.

- The **best Silhouette Score** of **0.657742** is achieved by the **Agglomerative** algorithm with the **Sqrt** normalization technique.
- The **best Calinski-Harabasz Index** of **15160.541299** is achieved by the **K-Means** algorithm with the **Sqrt** normalization technique.
- The **best Davies-Bouldin Index** of **0.46058** is achieved by both the **K-Means** and **Agglomerative** algorithms with the **YeoJohnson** normalization technique.